x509/src/certificate_document.rs

```diff
@@ -0,0 +1,97 @@
+//! CertificateDocument implementation
+
+use crate::Certificate;
+use der::{Error, Result};
+
+use alloc::vec::Vec;
+use core::fmt;
+use der::{Decodable, Document};
+
+#[cfg(feature = "pem")]
+use {core::str::FromStr, der::pem};
+
+/// Certificate document.
+///
+/// This type provides storage for [`Certificate`] encoded as ASN.1
+/// DER with the invariant that the contained-document is "well-formed", i.e.
+/// it will parse successfully according to this crate's parsing rules.
+#[derive(Clone)]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+pub struct CertificateDocument(Vec<u8>);
```

Comment:
@tarcieri I know I've implemented one of these document types for `pkcs10`. However:

1. What is the purpose of them? I don't understand how they are intended to be used. It seems that PEM decoding is implemented through the document types. However, in order to instantiate a document, you have to fully parse the DER. So why not just have the fully parsed type?

2. If there is a good reason for the document types, surely these can be genericized. They are all exactly the same code and differ only by the type of the value decoded therein. So `Document<T: Decodable + Encodable>` seems like an obvious win here.

Reply:
This seems like a complicated and important enough issue I did a long-form write up on it, so as not to lose the answers to these questions in a PR comment: #421

Some short answers though:

1. The purpose of these types is discussed at length in #421, but in regard to PEM specifically, since `Decode<'a>` is zero-copy and needs to borrow from a slice containing DER, parsing PEM needs to decode DER into a buffer that `Decode<'a>` can borrow from. If we had something like `DecodeOwned` which didn't need to borrow, documents could be decoded directly from PEM. As it were, this is how the `ssh-key` crate works, and I just added support for incremental/buffered PEM decoding to the `pem-rfc7468` crate (#406) and changed the `ssh-key` crate to use that (#407)

2. The problem with `Document<T: Decodable + Encodable>` is you omitted the lifetime on `Decodable<'a>`. Since it's a `'self` borrow and [Rust doesn't actually have a way to notate a `'self` lifetime](https://users.rust-lang.org/t/access-to-implicit-lifetime-of-containing-object-aka-self-lifetime/18917/17) you can't notate that on a struct definition, however as I noted in #421 it can probably be done by moving the bounds to individual methods so `Decodable<'a>` can hang off `pub fn decode<'a>(&'a self) -> T where T: Decode<'a>`